docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+
+## What is the expected outcome?
+
+Systems will run smoothly with a higher number of partitions than
+they could before, as long as some of those partitions are idle.
+
+# Guide-level explanation
+
+## Introducing new named concepts.
+
+`Hibernated topic`: a topic whose partitions do not have active raft
+groups running in the cluster.  It cannot serve user I/O (aside from
+metadata requests) until it is woken up.
+
+`Waking a topic`: a control plane operation to change the status of
+the topic, followed by raft elections for all partitions within the
+topic.
+
+These concepts are not visible to applications.  They are optionally visible
+to system administrators optionally:
+- system monitoring telemetry, where we would surface indicators for how many partitions
+  and topics are currently hibernated.
+- tunable configuration properties, where they could control the delay til hibernation
+  if needed.
+
+
+# Reference-level explanation
+
+## High level design
+
+The topic table, which stores the definition of the topics & partitions
+on the system, is driven by controller log commands.  New commands
+are added for hibernate and wake.  Nodes in a partition's raft group
+notice the hibernation of the topic, and shut down the raft group.
+
+There is a "hibernating" intermediate state where members of the
+partitions' raft groups are responsible for refusing writes, and
+reporting the highest offset back to the controller so that it can
+reliably store the frozen offset state of the topic when storing
+the controller message that finalizes the hibernation.
+
+Once the topic is hibernated, all its partitions will report a
+fictional leader, so that clients know where to send metadata requests,
+and so that monitoring systems do not incorrectly flag the partition
+as being in a problematic state.
+
+On startup, a hibernated partition's storage resources are not
+recovered or replayed.  Hibernated partitions are not subject to
+continuous retention policies, BUT the controller can notice if
+a partition has been hibernated for longer than its retention.ms
+limit, and wake it up for long enough to do housekeeping.
+
+We should still be able to serve metadata requests for idle topics,
+because otherwise any monitoring system will end up re-awakening
+the topics whenever querying them.
+
+## How do we decide when to hibernate?
+
+We don't persist an `atime` type stat for partitions, and rightly so,
+because otherwise we'd be writing metadata every time someone did
+a read.
+
+Detecting an idle partition at runtime therefore relies on a partition
+staying up with stable leadership long enough to reach the idle timeout.
+That's fine if the timeout is a few hours long (if we're changing leaders
+more often than that then we've got bigger problems) but more problematic
+if we have a long idle timeout (perhaps a day or more) with a reasonable
+chance of leadership changes or node restarts before we reach it.
```

Comment:
Can this be "solved" by classifying some operations as not affecting the determination of idleness (e.g. anything other than an application read or write)?

Reply:
If we could assume that only writes count, that would make the problem go away, but I'm pretty sure we need to consider reads too.

However... we could have a different threshold for writes and reads.  Like, to hibernate, a topic must have had no writes for 24h (easy to check from existing persistent state) and then also no reads (tracked with ephemeral state on leader) for like an hour.  Not necessarily exposing that complexity in the user config, maybe the write-side check is what they get to configure, and the read-side check is something hidden.

Reply:
> but I'm pretty sure we need to consider reads too.

I agree about reads. I think this was mentioned elsewhere in the RFC regarding atime. A very course grained recording in kvstore (or wherever we are keeping some per-group state) would be sufficient. To account for leadership changes, we could add a new rpc or find a handful of extra bits in an RPC to propagate atime.